claim/claim.go

```diff
@@ -151,3 +150,32 @@ func (c Claim) Validate() error {
 	// validate the Result
 	return errors.Wrap(c.Result.Validate(), "claim validation failed")
 }
+
+// ulidMutex guards the generation of ULIDs, because the use of rand
+// is not thread-safe.
+var ulidMutex sync.Mutex
+
+// ulidEntropy must be set once and reused when generating ULIDs, to guarantee
+// that each ULID is monotonically increasing.
+var ulidEntropy = ulid.Monotonic(rand.New(rand.NewSource(time.Now().UnixNano())), 0)
+
+// MustNewULID generates a string representation of a ULID and panics on failure
+// instead of returning an error.
+func MustNewULID() string {
+	result, err := NewULID()
+	if err != nil {
+		panic(err)
+	}
+	return result
+}
+
+// NewULID generates a string representation of a ULID.
+func NewULID() (string, error) {
+	ulidMutex.Lock()
```

Comment:
I'm a bit concerned with locking everything everytime we update or create a new Claim.
The ULID documentation says the following:

> Please note that rand.Rand from the math package is not safe for concurrent use. Instantiate one per long living go-routine or use a sync.Pool if you want to avoid the potential contention of a locked rand.Source as its been frequently observed in the package level functions.

What about creating a new instance of rand and entropy each time we need it instead?
It may not make sense from a cryptographic point of view, but I think it's not the point here (otherwise we would use cryptographic rand instead?)
What about the following ?
```go
func NewULID() (string, error) {
    	now := time.Now()
	entropy := ulid.Monotonic(rand.New(rand.NewSource(now.UnixNano())), 0)
        result, err := ulid.New(ulid.Timestamp(now), entropy)
  	if err != nil {
		return "", errors.Wrap(err, "could not generate a new ULID")
	}
        return result.String(), nil
}
```

Reply:
That doesn't yield increasing ulids unfortunately when called successively within small timeframes. You have to keep the `entropy` variable and reuse it between calls to `ulid.New` in order to have that work. Without this, unit tests that generate ulids back to back fail to sort properly. In a batch system (not a CLI with a user typing behind it), this could also be a problem.

I just took your implementation and ran it through `TestNewULID_ThreadSafe` which also asserts that each new ulid increases and it fails.

Below are some relevant issues from the ulid library. If you see a better recommended implementation, maybe you can submit a fix for what we have? Because what we are using is borked. 😀 

https://github.com/oklog/ulid/issues/8
https://github.com/oklog/ulid/issues/33

Reply:
> That doesn't yield increasing ulids unfortunately when called successively within small timeframes.

Indeed 🤦 

> Because what we are using is borked. 😀

Well I guess your implementation is then the less worst option 😄 But I think we then need make sure that the `NewULID` and `MustNewULID` function documentations stated explicitly that they are thread safe and use a global mutex.

Reply:
Haha, yeah that's how I felt prototyping out solutions. "This is the least worst" 😀 

I'll update the documentation to include that explanation.